File: builder/app/src/lib/catalog/v2/grid.ts
// The OBCA cell grid, mirrored in TypeScript.
//
// This is a *mirror*, not a second design: every constant, every rounding
// direction and every half-open edge here is normative in `OBCA_Spec.md` §1 and
// implemented on the producer side in `host/obc-pack/src/grid.rs`. The builder
// needs the same arithmetic — a box drawn on the map has to resolve to exactly
// the cell ids the bakery published, and a coverage outline has to be drawn from
// exactly the squares those ids name — so the two implementations must agree to
// the microdegree. `grid.test.ts` therefore pins the *same concrete vectors* the
// Rust tests pin, and that pinning is the only thing standing between us and a
// silent drift that shows up as a one-cell hole in someone's map.
//
// Two conventions differ from the Rust side on purpose:
//
//   * Rust's `UBox` is `(min_lon, min_lat, max_lon, max_lat)` because it is
//     handed straight to the serializer. Nothing here talks to a serializer, so
//     a box is a named object in `lat, lon` order — the order the catalog, the
//     OBCM header and every boundary ring use.
//   * Wire shapes in this app keep their snake_case field names (they are the
//     document). A box computed here is not a wire shape, so it is camelCase.
//
// Everything is integer microdegrees and every value stays far inside the range
// a double represents exactly: the world box is ±2^28 µdeg and the largest
// intermediate is 2^29, against a 2^53 exact-integer limit. No coordinate in
// this module is ever a float.

/** Origin of the fixed global cell grid, µdeg, on **both** axes (§1.1).
 *
 *  A power of two, so every permitted cell size divides it exactly — which is
 *  the whole reason the grid is not anchored at −90/−180. */
export const GRID_ORIGIN = -268_435_456;

/** Side of the world box, µdeg (§1.1): `2^29`, i.e. ≈ ±268.435456°. Strictly
 *  larger than the geographic domain, and the grid does **not** wrap. */
export const WORLD_SIDE = 536_870_912;

/** Smallest permitted cell size as `log2(µdeg)` (§1.1). */
export const MIN_CELL_LOG2 = 10;

/** Largest permitted cell size as `log2(µdeg)` (§1.1). */
export const MAX_CELL_LOG2 = 28;

/** A cell id, or a coordinate, that the grid does not admit. Thrown rather than
 *  returned as a null: every caller here is working from a catalog document that
 *  a producer already validated, so a failure is a bug or a corrupt document,
 *  and neither should be papered over with a fallback cell. */
export class GridError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "GridError";
    }
}

/** One cell of the grid: a size (as `log2(µdeg)`) and its **latitude** index `i`
 *  and **longitude** index `j` — the canonical id is `<log2>/<i>/<j>`, latitude
 *  first (§1.3). */
export interface CellId {
    readonly log2: number;
    readonly i: number;
    readonly j: number;
}

/** A box in integer microdegrees. Half-open on `max` wherever a cell square is
 *  meant (§1.1); an ordinary bbox from a user's drag is closed, and
 *  {@link cellsIntersecting} is written to make both behave the same way. */
export interface UBox {
    minLat: number;
    minLon: number;
    maxLat: number;
    maxLon: number;
}

/** Cell size in µdeg. */
export function cellSize(log2: number): number {
    return 2 ** log2;
}

/** Cells per axis at size `2^log2`. */
export function axisCells(log2: number): number {
    return WORLD_SIDE / 2 ** log2;
}

/** Zero-padding width of a cell id's indices (§1.3): `max(4, digits(cells per
 *  axis − 1))`. Four for every size at or above `2^16`, wider below — producers
 *  MUST widen rather than truncate, and a consumer that assumed four would fail
 *  to key a `2^10` id against its own index. */
export function idWidth(log2: number): number {
    return Math.max(4, String(axisCells(log2) - 1).length);
}

function checkLog2(log2: number): void {
    if (!Number.isInteger(log2) || log2 < MIN_CELL_LOG2 || log2 > MAX_CELL_LOG2) {
        throw new GridError(
            `cell size 2^${log2} µdeg is outside the grid's ${MIN_CELL_LOG2}..=${MAX_CELL_LOG2}`,
        );
    }
}

/** A cell by size + indices, validated against the world box. */
export function cellId(log2: number, i: number, j: number): CellId {
    checkLog2(log2);
    const n = axisCells(log2);
    const bad = (v: number) => !Number.isInteger(v) || v < 0 || v >= n;
    if (bad(i) || bad(j)) {
        throw new GridError(
            `cell 2^${log2}/${i}/${j} is outside the world box (indices must be 0..${n})`,
        );
    }
    return { log2, i, j };
}

/** Parse a canonical id `<log2>/<i>/<j>` (§1.3). Lenient about zero padding on
 *  the way in, strict on the way out ({@link formatCellId} pads canonically) —
 *  the same asymmetry the Rust side has, so an id written by either round-trips
 *  through the other. */
export function parseCellId(s: string): CellId {
    const parts = s.split("/");
    if (parts.length !== 3 || parts.some((p) => !/^\d+$/.test(p))) {
        throw new GridError(`cell id ${JSON.stringify(s)} is not <log2>/<i>/<j>`);
    }
    return cellId(Number(parts[0]), Number(parts[1]), Number(parts[2]));
}

/** The canonical, zero-padded id (§1.3). Also this app's map key for a cell:
 *  canonical means two spellings of one cell cannot end up as two entries. */
export function formatCellId(cell: CellId): string {
    const w = idWidth(cell.log2);
    return `${cell.log2}/${String(cell.i).padStart(w, "0")}/${String(cell.j).padStart(w, "0")}`;
}

/** The cell's square, half-open on both axes (§1.1).
 *
 *  This is the whole of a cell's coverage: a catalog entry carries no bbox
 *  precisely because the id determines the square to the microdegree, and the
 *  bakery verifies the artifact's own OBCM header against it. */
export function cellSquare(cell: CellId): UBox {
    const s = cellSize(cell.log2);
    const minLat = GRID_ORIGIN + cell.i * s;
    const minLon = GRID_ORIGIN + cell.j * s;
    return { minLat, minLon, maxLat: minLat + s, maxLon: minLon + s };
}

/** Floor division — `Math.trunc` would round toward zero and drift a whole cell
 *  south/west of the origin, which is exactly the bug the negative-origin tests
 *  exist to catch. */
function floorDiv(a: number, b: number): number {
    return Math.floor(a / b);
}

/** The cell of size `2^log2` whose half-open square contains `(lat, lon)`.
 *
 *  Half-open is the point: a coordinate exactly on a cell's `max` edge belongs
 *  to the **next** cell, so a point is owned by exactly one cell of a size. */
export function cellContaining(log2: number, lat: number, lon: number): CellId {
    return {
        log2,
        i: floorDiv(lat - GRID_ORIGIN, cellSize(log2)),
        j: floorDiv(lon - GRID_ORIGIN, cellSize(log2)),
    };
}

/** Whether `(lat, lon)` lies in this cell's half-open square. */
export function cellContains(cell: CellId, lat: number, lon: number): boolean {
    const { minLat, minLon, maxLat, maxLon } = cellSquare(cell);
    return lat >= minLat && lat < maxLat && lon >= minLon && lon < maxLon;
}

/**
 * Every cell of size `2^log2` whose square intersects `box`, in ascending
 * `(i, j)` order.
 *
 * Intersection is decided on the **half-open** squares, so the `max` edges of
 * `box` are inclusive of the cell that owns them: a vertex sitting exactly on a
 * grid line belongs to the cell above / east of it, and that cell is therefore
 * part of the covering. Cells outside the world box are clamped away rather than
 * wrapped (§1.4).
 *
 * This is OBCA §1.2's coverage rule in one function, and the *generous coarse*
 * behaviour the epic wants falls straight out of it: run it per band and a
 * corridor is covered precisely at `2^18` and generously — whole covering cells,
 * i.e. context beyond the selection — at `2^20`. There is no second rule and no
 * special case; the generosity is a consequence of cell size.
 */
export function cellsIntersecting(log2: number, box: UBox): CellId[] {
    checkLog2(log2);
    if (box.minLon > box.maxLon || box.minLat > box.maxLat) return [];
    const s = cellSize(log2);
    const n = axisCells(log2);
    const index = (v: number) => Math.min(Math.max(floorDiv(v - GRID_ORIGIN, s), 0), n - 1);
    const i0 = index(box.minLat);
    const i1 = index(box.maxLat);
    const j0 = index(box.minLon);
    const j1 = index(box.maxLon);
    const out: CellId[] = [];
    for (let i = i0; i <= i1; i++) {
        for (let j = j0; j <= j1; j++) out.push({ log2, i, j });
    }
    return out;
}

/** Whether `v` lies exactly on a grid line of size `2^log2` — i.e. on a cell
 *  boundary. A pure function of the coordinate, which is why two neighbours
 *  cannot disagree about it (§3.4). */
export function onGridLine(v: number, log2: number): boolean {
    const s = cellSize(log2);
    return (v - GRID_ORIGIN) % s === 0;
}

/**
 * The bounding box of a set of cells' squares.
 *
 * Not the coverage outline — that is the union of the squares and is drawn as
 * its true stair-edged shape (`OBCC_Spec.md` §11.8). This is only what a map
 * view has to fit, and it is `null` for an empty set rather than a degenerate
 * box at the origin.
 */
export function coverageBbox(cells: Iterable<CellId>): UBox | null {
    let box: UBox | null = null;
    for (const cell of cells) {
        const sq = cellSquare(cell);
        if (!box) {
            box = { ...sq };
            continue;
        }
        box.minLat = Math.min(box.minLat, sq.minLat);
        box.minLon = Math.min(box.minLon, sq.minLon);
        box.maxLat = Math.max(box.maxLat, sq.maxLat);
        box.maxLon = Math.max(box.maxLon, sq.maxLon);
    }
    return box;
}

File: firmware/obc-elevation/src/grid.rs
//! The sample lattice: integer arithmetic from a µdeg coordinate to a lattice index, a cell index,
//! and a tile-local sample (`OBCT_Spec.md` §1, §3).
//!
//! Deliberately **integer-only and float-free**, for the same reason OBCA's grid is: the whole
//! contract rests on two implementations landing on the same sample from the same coordinate, and a
//! float would make that a property of the FPU. Everything here is a shift or a mask over an `i64`
//! widened from the `int32` µdeg coordinate — no division, no rounding, no `libm`.

use obc_formats::obct::{GRID_ORIGIN, WORLD_SIDE};

/// Where a query coordinate lands on the lattice: the index of the sample at or below it on each
/// axis, plus the sub-posting remainder that becomes the bilinear weight (`OBCT_Spec.md` §5.1).
///
/// `frac_*` is in µdeg, `0..P`, **not** a normalized fraction — keeping it in the coordinate's own
/// unit is what lets §5.2's interpolation stay in integers.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Lattice {
    /// Latitude sample index: `floor((lat − GRID_ORIGIN) / P)`.
    pub i: u32,
    /// Longitude sample index: `floor((lon − GRID_ORIGIN) / P)`.
    pub j: u32,
    /// `lat − lattice_lat(i)`, µdeg, in `0..P`.
    pub frac_lat: u32,
    /// `lon − lattice_lon(j)`, µdeg, in `0..P`.
    pub frac_lon: u32,
}

/// Locate `(lat, lon)` on the lattice of posting `2^posting_log2`, or `None` when the coordinate is
/// outside the world box (`OBCT_Spec.md` §1.1). The world box is bigger than the geographic domain,
/// so this rejects only genuinely impossible coordinates — a `0` from an unfixed GPS is *inside* it
/// and is the sampler's business to answer, not this function's.
#[inline]
pub fn locate(lat_udeg: i32, lon_udeg: i32, posting_log2: u8) -> Option<Lattice> {
    let (i, frac_lat) = axis(lat_udeg, posting_log2)?;
    let (j, frac_lon) = axis(lon_udeg, posting_log2)?;
    Some(Lattice { i, j, frac_lat, frac_lon })
}

/// One axis of [`locate`]. Widened to `i64` before the subtraction: `lat − GRID_ORIGIN` adds `2^28`
/// to a value that may legally be any `int32`, which overflows `i32` near its top end.
#[inline]
fn axis(v_udeg: i32, posting_log2: u8) -> Option<(u32, u32)> {
    let offset = v_udeg as i64 - GRID_ORIGIN as i64;
    if offset < 0 || offset >= WORLD_SIDE as i64 {
        return None;
    }
    let mask = (1i64 << posting_log2) - 1;
    Some(((offset >> posting_log2) as u32, (offset & mask) as u32))
}

/// The µdeg coordinate of lattice index `i` on either axis: `GRID_ORIGIN + i·P`.
#[inline]
pub fn lattice_coord(i: u32, posting_log2: u8) -> i32 {
    (GRID_ORIGIN as i64 + ((i as i64) << posting_log2)) as i32
}

/// The cell index owning lattice index `i`, for a cell of `2^cell_log2` µdeg at this posting.
/// A pure shift, which is exactly why the cell side is a power of two on the same origin: cell
/// membership is derivable from the lattice index with no table and no search.
#[inline]
pub fn cell_of(i: u32, posting_log2: u8, cell_log2: u8) -> u32 {
    i >> (cell_log2 - posting_log2)
}

/// The lattice index of a cell's **min** sample on either axis — the sample the cell owns under the
/// half-open rule (`OBCT_Spec.md` §3.1).
#[inline]
pub fn cell_base_sample(cell: u32, posting_log2: u8, cell_log2: u8) -> u32 {
    cell << (cell_log2 - posting_log2)
}

/// Number of cells along one axis of the world box at `2^cell_log2` µdeg (OBCA §1.1).
#[inline]
pub fn axis_cells(cell_log2: u8) -> u32 {
    WORLD_SIDE >> cell_log2
}

/// Zero-padding width of a cell index in a canonical cell id (`OBCA_Spec.md` §1.3):
/// `max(4, digits(axis_cells − 1))`. Four digits at `2^16` and above, wider below.
///
/// The rule is here, in the one crate both sides of the terrain pipeline already depend on,
/// because it is **content addressing**, not formatting: `18/1204/52` and `18/01204/1052` are two
/// strings for one cell, and a store keyed by the string would then hold the same square twice
/// under two names. `obc-pack`'s `grid::id_width` and `obc-dem`'s `bake::cell_file_name` are the
/// two producers of those strings, they cannot see each other (a host tool must not depend on the
/// packer), and a second transcription of `max(4, …)` is exactly how the two would drift apart.
///
/// Integer-only and allocation-free, like everything else in this crate — the digit count is a
/// loop, not a `to_string().len()`.
pub fn id_width(cell_log2: u8) -> usize {
    let mut v = axis_cells(cell_log2) - 1;
    let mut digits = 1;
    while v >= 10 {
        v /= 10;
        digits += 1;
    }
    if digits < 4 {
        4
    } else {
        digits
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The lattice is anchored on the OBCA origin, not on the query — the same coordinate must land
    /// on the same index whatever file asks.
    #[test]
    fn the_origin_is_lattice_index_zero_on_both_axes() {
        let at_origin = locate(GRID_ORIGIN, GRID_ORIGIN, 9).unwrap();
        assert_eq!(at_origin, Lattice { i: 0, j: 0, frac_lat: 0, frac_lon: 0 });
        assert_eq!(lattice_coord(0, 9), GRID_ORIGIN);
        // One posting in, on latitude only.
        let one = locate(GRID_ORIGIN + 512, GRID_ORIGIN, 9).unwrap();
        assert_eq!((one.i, one.j, one.frac_lat), (1, 0, 0));
        assert_eq!(lattice_coord(1, 9), GRID_ORIGIN + 512);
    }

    #[test]
    fn a_coordinate_splits_into_index_and_sub_posting_remainder() {
        // 47°N / 8°E at the v1 posting.
        let l = locate(47_000_000, 8_000_000, 9).unwrap();
        assert_eq!(l.i, ((47_000_000i64 + 268_435_456) >> 9) as u32);
        assert_eq!(lattice_coord(l.i, 9) as i64 + l.frac_lat as i64, 47_000_000);
        assert_eq!(lattice_coord(l.j, 9) as i64 + l.frac_lon as i64, 8_000_000);
        assert!(l.frac_lat < 512 && l.frac_lon < 512);
    }

    #[test]
    fn the_world_box_is_half_open_and_int32_safe() {
        assert!(locate(GRID_ORIGIN - 1, 0, 9).is_none());
        assert!(locate(0, GRID_ORIGIN - 1, 9).is_none());
        let last = GRID_ORIGIN + WORLD_SIDE as i32 - 1;
        assert!(locate(last, last, 9).is_some());
        assert!(locate(i32::MAX, 0, 9).is_none(), "the widening keeps this a rejection, not a wrap");
        assert!(locate(i32::MIN, 0, 9).is_none());
    }

    /// Cell membership and the lattice must agree in both directions: the base sample of the cell a
    /// sample belongs to is never above that sample, and never a whole cell below it.
    #[test]
    fn cell_membership_agrees_with_the_lattice_at_every_v1_pairing() {
        let (posting_log2, cell_log2) = (9u8, 19u8);
        let span = 1u32 << (cell_log2 - posting_log2);
        for i in [0u32, 1, span - 1, span, span + 1, 12_345, u32::MAX >> cell_log2] {
            let c = cell_of(i, posting_log2, cell_log2);
            let base = cell_base_sample(c, posting_log2, cell_log2);
            assert!(base <= i && i - base < span, "sample {i} sits in cell {c} at base {base}");
            assert_eq!(cell_of(base, posting_log2, cell_log2), c, "a cell's base sample is its own");
        }
        assert_eq!(axis_cells(19), 1 << 10);
        assert_eq!(axis_cells(28), 2, "the largest legal cell tiles the world box twice per axis");
    }
}
